Mathlib/AlgebraicTopology/SimplicialSet/StrictSegal.lean

```diff
@@ -0,0 +1,156 @@
+  (Equiv.ofBijective _ (segal n)).invFun
+
+@[simp]
+theorem spineToSimplex_spine (f : Path X n) :
+    X.spine n (StrictSegal.spineToSimplex f) = f :=
+  (Equiv.ofBijective _ (segal n)).right_inv f
+
+@[simp]
+theorem spineToSimplex_vertex (i : Fin (n + 1)) (f : Path X n) :
+    X.map (const [0] [n] i).op (spineToSimplex f) = f.vertex i := by
+  rw [← spine_vertex, spineToSimplex_spine]
+
+@[simp]
+theorem spineToSimplex_spine_edge (i : Fin n) (f : Path X n) :
+    X.map (mkOfSucc i).op (spineToSimplex f) = f.arrow i := by
+  rw [← spine_arrow, spineToSimplex_spine]
+
+
+/-- In the presence of the strict Segal condition, a path of length `n` can be "composed" by taking
+the diagonal edge of the resulting `n`-simplex. -/
+noncomputable def spineToDiagonal (f : Path X n) : X _[1] := diagonal (spineToSimplex f)
+
+@[simp]
+theorem spineToSimplex_interval (f : Path X n) (j l : ℕ) (hjl : j + l < n + 1)  :
+    X.map (subinterval j l hjl).op (spineToSimplex f) =
+      spineToSimplex (Path.interval f j l hjl) := by
+  apply (segal _).injective
+  rw [StrictSegal.spineToSimplex_spine]
+  ext i
+  · simp only [mkHom, Equiv.invFun_as_coe, spine_vertex, Fin.coe_addNat,
+      ← FunctorToTypes.map_comp_apply, ← op_comp, const_comp, len_mk, spineToSimplex_vertex]
+    congr
+    exact Fin.eq_of_val_eq (Nat.mod_eq_of_lt (by omega)).symm
+  · unfold Path.interval
+    simp only [Equiv.invFun_as_coe, spine_arrow, Fin.coe_addNat]
+    simp only [← FunctorToTypes.map_comp_apply, ← op_comp]
+    have ceq : mkOfSucc i ≫ subinterval j l hjl = mkOfSucc ⟨i + j, (by omega)⟩ := by
+      ext ⟨e, he⟩ : 3
+      unfold subinterval
+      match e with
+      | 0 => rfl
+      | 1 => ?_
+      conv_rhs =>
+        apply mkOfSucc_homToOrderHom_one
+      simp only [len_mk, Nat.reduceAdd, mkHom, comp_toOrderHom, Hom.toOrderHom_mk, Fin.mk_one,
+        Fin.isValue, OrderHom.comp_coe, OrderHom.coe_mk, Function.comp_apply, Fin.succ_mk,
+        Fin.mk.injEq]
+      exact Nat.succ_add_eq_add_succ ↑i ↑j
+    rw [ceq]
+    simp only [spineToSimplex_spine_edge]
+
+@[simp]
+theorem spineToSimplex_edge (f : Path X n) (j l : ℕ) (hn : j + l < n + 1) :
+    X.map (mkOfLe ⟨j, (by omega)⟩ ⟨j + l, hn⟩ (Nat.le_add_right j l)).op (spineToSimplex f) =
+      spineToDiagonal (Path.interval f j l hn) := by
+  unfold spineToDiagonal
+  rw [← congrArg diagonal (spineToSimplex_interval f j l hn)]
+  unfold diagonal
+  simp only [← FunctorToTypes.map_comp_apply, ← op_comp]
+  have : mkOfLe ⟨j, (by omega)⟩ ⟨j + l, hn⟩ (Nat.le_add_right j l) =
+      mkOfDiag l ≫ subinterval j l hn := by
+    ext e : 3
+    unfold subinterval mkOfDiag mkOfLe
+    simp only [len_mk, Nat.reduceAdd, mkHom, Hom.toOrderHom_mk, OrderHom.coe_mk,
+      Fin.natCast_eq_last, comp_toOrderHom, OrderHom.mk_comp_mk, Function.comp_apply]
+    match e with
+    | 0 => simp
+    | 1 =>
+    apply Fin.eq_of_val_eq
+    simp only [Fin.isValue, Fin.val_last]
+    exact Nat.add_comm j l
+  rw [this]
```

Comment:
If you think it's natural, from a readability pov, you could restructure this as
```suggestion
  suffices mkOfLe ⟨j, (by omega)⟩ ⟨j + l, hn⟩ (Nat.le_add_right j l) =
      mkOfDiag l ≫ subinterval j l hn by
    rw [this]
  ext e : 3
  unfold subinterval mkOfDiag mkOfLe
  simp only [len_mk, Nat.reduceAdd, mkHom, Hom.toOrderHom_mk, OrderHom.coe_mk,
    Fin.natCast_eq_last, comp_toOrderHom, OrderHom.mk_comp_mk, Function.comp_apply]
  match e with
  | 0 => simp
  | 1 =>
  apply Fin.eq_of_val_eq
  simp only [Fin.isValue, Fin.val_last]
  exact Nat.add_comm j l
```

Reply:
Done (in the corresponding lemma).